agent/src/util/device_plugin_service.rs

```diff
@@ -403,42 +802,160 @@ impl InstanceDevicePlugin {
     }
 }
 
-/// This returns true if this node can reserve a `device_usage_id` slot for an instance
-/// and false if it is already reserved.
-/// # More details
-/// Cases based on the usage slot (`device_usage_id`) value
-/// 1. device_usage\[id\] == "" ... this means that the device is available for use
-///     * (ACTION) return true
-/// 2. device_usage\[id\] == self.nodeName ... this means THIS node previously used id, but the DevicePluginManager knows that this is no longer true
-///     * (ACTION) return false
-/// 3. device_usage\[id\] == (some other node) ... this means that we believe this device is in use by another node and should be marked unhealthy
-///     * (ACTION) return error
-/// 4. No corresponding id found ... this is an unknown error condition (BAD)
-///     * (ACTION) return error
-fn slot_available_to_reserve(
-    device_usage_id: &str,
-    node_name: &str,
-    instance: &InstanceSpec,
-) -> Result<bool, Status> {
-    if let Some(allocated_node) = instance.device_usage.get(device_usage_id) {
-        if allocated_node.is_empty() {
-            Ok(true)
-        } else if allocated_node == node_name {
-            Ok(false)
-        } else {
-            trace!("slot_available_to_reserve - request for device slot {} previously claimed by a diff node {} than this one {} ... indicates the device on THIS node must be marked unhealthy, invoking ListAndWatch ... returning failure, next scheduling should succeed!", device_usage_id, allocated_node, node_name);
-            Err(Status::new(
-                Code::Unknown,
-                "Requested device already in use",
-            ))
-        }
-    } else {
-        // No corresponding id found
-        trace!(
-            "slot_available_to_reserve - could not find {} id in device_usage",
-            device_usage_id
-        );
-        Err(Status::new(
+fn build_virtual_devices_list_for_instance(
+    device_map: HashMap<String, Vec<v1beta1::Device>>,
```

Comment:
This should be able to take in a reference to avoid unnecessary cloning
```suggestion
    device_map: &HashMap<String, Vec<v1beta1::Device>>,
```

Reply:
pass reference instead of value